### PR TITLE
Support ChangedFiles for Github & Gitlab PRs and Gitlab pushes

### LINKED ISSUE #555: Lists of changed files of pull-requests are empty
### Component

server

### Describe the bug

```yml
pipeline:
  build:
    image: node:14-alpine
    commands:
      - cd docs/
      - yarn install --frozen-lockfile
      - yarn build
    when:
      event: [push, pull_request]
      path: "docs/**"
```

Step is executed although no file in `docs/**` has been changed.

### System Info

```shell
{"source":"https://github.com/woodpecker-ci/woodpecker","version":"next-7931bc78"}
```

### Additional context

```shell
https://ci.woodpecker-ci.org/woodpecker-ci/woodpecker/build/1317
```

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy).

### PR DESCRIPTION
**Pull request by anbraten**
closes #555 

load changed files for Github and Gitlab PRs
Gitea is waiting for https://github.com/go-gitea/gitea/pull/18228

TODOs:
- [x] test gitlab in real
- [x] update gitlab webhook fixtures

### CODE REVIEW

File: server/remote/github/github.go [hunk 1]
```diff
@@ -491,6 +492,51 @@ func (c *client) Branches(ctx context.Context, u *model.User, r *model.Repo) ([]
...
+func (c *client) loadChangedFilesFromPullRequest(ctx context.Context, repo *model.Repo, build *model.Build) (*model.Build, error) {
+	user, err := store.FromContext(ctx).GetUser(repo.UserID)
+	if err != nil {
+		return nil, err
+	}
+
+	client := c.newClientToken(ctx, user.Token)
+
+	// extract id from ref: "refs/pull/%d/merge"
+	pullRequestID, err := strconv.Atoi(build.Ref)
```
Comment (6543) [inline server/remote/github/github.go:520]: Can we use:

fmt.Sscanf(build.Ref, "refs/pull/%d/merge", ...

File: server/api/hook.go [hunk 2]
```diff
@@ -311,6 +312,12 @@ func zeroSteps(build *model.Build, remoteYamlConfigs []*remote.FileMeta) bool {
 		return true
 	}
 
+	for _, buildItem := range buildItems {
+		if buildItem.Proc.State != model.StatusPending {
```
Comment (6543) [inline server/api/hook.go:316]: Will have no efect!

File: server/api/hook.go [hunk 3]
```diff
@@ -311,6 +312,12 @@ func zeroSteps(build *model.Build, remoteYamlConfigs []*remote.FileMeta) bool {
...
+			return false
```
Comment (6543) [inline server/api/hook.go:317]: this change is also unrelated

File: server/remote/github/github.go [hunk 4]
```diff
@@ -491,6 +492,51 @@ func (c *client) Branches(ctx context.Context, u *model.User, r *model.Repo) ([]
...
+
+func (c *client) loadChangedFilesFromPullRequest(ctx context.Context, repo *model.Repo, build *model.Build) (*model.Build, error) {
+	user, err := store.FromContext(ctx).GetUser(repo.UserID)
+	if err != nil {
+		return nil, err
+	}
+
+	client := c.newClientToken(ctx, user.Token)
+
+	// extract id from ref: "refs/pull/%d/merge"
```
Comment (6543) [inline server/remote/github/github.go:519]: I have a rewrite of github webhook parsing via github lib in a new branch ... I'll let this as for merge and will improve it in mentioned upcomming pull

File: server/remote/gitlab/convert.go [hunk 5]
```diff
@@ -129,6 +129,14 @@ func convertMergeRequestHock(hook *gitlab.MergeEvent, req *http.Request) (*model
 	build.Title = obj.Title
 	build.Link = obj.URL
 
+	if hook.ObjectAttributes.StDiffs != nil {
+		files := make([]string, 0, len(hook.ObjectAttributes.StDiffs)*2)
```
Comment (6543) [inline server/remote/gitlab/convert.go:133]: need to be tested in real world

File: server/remote/gitlab/gitlab_test.go [hunk 6]
```diff
@@ -208,18 +210,19 @@ func Test_Gitlab(t *testing.T) {
 					req, _ := http.NewRequest(
 						testdata.ServiceHookMethod,
 						testdata.ServiceHookURL.String(),
-						bytes.NewReader(testdata.ServiceHookMergeRequestBody),
+						bytes.NewReader(testdata.ServiceHookMergeRequestBody), // TODO: update test data with new hooks
```
Comment (6543) [inline server/remote/gitlab/gitlab_test.go:213]: this todo should be done in this pull

File: server/remote/github/github.go (hunk 4 above)
Comment (6543) [reply server/remote/github/github.go:519]: #701 (based on this branch)

File: shared/utils/strings.go [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package utils
+
+// DedupStrings deduplicate string list, empty items are dropped
+func DedupStrings(list []string) []string {
```
Comment (6543) [inline shared/utils/strings.go:4]: need tests

File: server/remote/gitea/helper.go [hunk 8]
```diff
@@ -110,15 +111,14 @@ func buildFromPush(hook *pushHook) *model.Build {
 }
 
 func getChangedFilesFromPushHook(hook *pushHook) []string {
-	files := make([]string, 0)
-
+	files := make([]string, 0, len(hook.Commits)*4)
```
Comment (anbraten) [inline server/remote/gitea/helper.go:114]: Why `len(hook.Commits)*4`?

File: server/remote/github/github.go [hunk 9]
```diff
@@ -491,6 +493,53 @@ func (c *client) Branches(ctx context.Context, u *model.User, r *model.Repo) ([]
...
+		}
+	}
+
+	return repo, build, nil
+}
+
+func (c *client) loadChangedFilesFromPullRequest(ctx context.Context, pull *github.PullRequest, tmpRepo *model.Repo, build *model.Build) (*model.Build, error) {
+	_store, ok := store.TryFromContext(ctx)
+	if !ok {
+		return build, nil
```
Comment (anbraten) [inline server/remote/github/github.go:515]: We should handle this error.

File: server/remote/github/parse.go [hunk 10]
```diff
@@ -48,59 +48,142 @@ func parseHook(r *http.Request, merge bool) (*model.Repo, *model.Build, error) {
...
 
 // parsePullHook parses a pull request hook and returns the Repo and Build
 // details. If the pull request is closed nil values are returned.
-func parsePullHook(payload []byte, merge bool) (*model.Repo, *model.Build, error) {
-	hook := new(webhook)
-	err := json.Unmarshal(payload, hook)
-	if err != nil {
-		return nil, nil, err
+func parsePullHook(hook *github.PullRequestEvent, merge, privateMode bool) (*github.PullRequest, *model.Repo, *model.Build, error) {
+	// ignore these
```
Comment (anbraten) [inline server/remote/github/parse.go:149]: Could you improve this comment.

**Review by 6543 — APPROVED** [review]
ACK from my side

File: server/remote/gitlab/convert.go [hunk 11]
```diff
@@ -129,10 +133,10 @@ func convertMergeRequestHock(hook *gitlab.MergeEvent, req *http.Request) (*model
 	build.Title = obj.Title
 	build.Link = obj.URL
 
-	return repo, build, nil
+	return obj.IID, repo, build, nil
```
Comment (anbraten) [inline server/remote/gitlab/convert.go:136]: Is this `obj.IID` really correct?

File: server/remote/gitlab/convert.go [hunk 12]
```diff
@@ -161,6 +165,7 @@ func convertPushHock(hook *gitlab.PushEvent) (*model.Repo, *model.Build, error)
 	build.Branch = strings.TrimPrefix(hook.Ref, "refs/heads/")
 	build.Ref = hook.Ref
 
+	files := make([]string, 0, len(hook.Commits)*4)
```
Comment (anbraten) [inline server/remote/gitlab/convert.go:168]: Why do you used `len(hook.Commits)*4` as capacity?

File: server/remote/gitlab/gitlab.go [hunk 13]
```diff
@@ -538,12 +540,61 @@ func (g *Gitlab) Hook(req *http.Request) (*model.Repo, *model.Build, error) {
...
+		return convertTagHook(event)
 	default:
 		return nil, nil, nil
 	}
 }
+
+func (g *Gitlab) loadChangedFilesFromMergeRequest(ctx context.Context, tmpRepo *model.Repo, build *model.Build, mergeIID int) (*model.Build, error) {
+	_store, ok := store.TryFromContext(ctx)
+	if !ok {
+		return build, nil
```
Comment (anbraten) [inline server/remote/gitlab/gitlab.go:565]: We should handle this error.

File: server/remote/gitea/helper.go (hunk 8 above)
Comment (6543) [reply server/remote/gitea/helper.go:114]: to make it faster (less re-alocations), if you guess ca. the size you need it's better.

the idea behind this number is: "a normal number of changed files per commit = 4"

File: server/remote/github/github.go (hunk 9 above)
Comment (6543) [reply server/remote/github/github.go:515]: how do you like to handle it, to be specific?

File: server/remote/gitlab/convert.go (hunk 11 above)
Comment (6543) [reply server/remote/gitlab/convert.go:136]: that's wat you use in your ref too ...

File: server/remote/gitlab/convert.go (hunk 12 above)
Comment (6543) [reply server/remote/gitlab/convert.go:168]: as https://github.com/woodpecker-ci/woodpecker/pull/697#discussion_r786300852

File: server/remote/github/github.go (hunk 9 above)
Comment (anbraten) [reply server/remote/github/github.go:515]: log it?

File: server/remote/github/parse.go (hunk 10 above)
Comment (6543) [reply server/remote/github/parse.go:149]:
```suggestion
	// ignore all pull hooks beside one of a new pull and push updates of open one
```

Comment (6543) [reply server/remote/github/parse.go:149]: ?

File: server/remote/gitea/helper.go (hunk 8 above)
Comment (6543) [reply server/remote/gitea/helper.go:114]: PS: if it do not match the exact size it still works as intended

File: server/remote/gitlab/convert.go (hunk 12 above)
Comment (anbraten) [reply server/remote/gitlab/convert.go:168]:
```suggestion
        // assume a capacity of 4 changed files per commit
	files := make([]string, 0, len(hook.Commits)*4)
```

File: server/remote/gitea/helper.go (hunk 8 above)
Comment (anbraten) [reply server/remote/gitea/helper.go:114]:
```suggestion
  // assume a capacity of 4 changed files per commit
	files := make([]string, 0, len(hook.Commits)*4)
```

File: server/remote/gitlab/convert.go [hunk 14]
```diff
@@ -68,17 +73,17 @@ func convertMergeRequestHock(hook *gitlab.MergeEvent, req *http.Request) (*model
 	obj := hook.ObjectAttributes
 
 	if target == nil && source == nil {
-		return nil, nil, fmt.Errorf("target and source keys expected in merge request hook")
+		return 0, nil, nil, fmt.Errorf("target and source keys expected in merge request hook")
```
[the remark this answers is not in the document]
Comment (6543) [reply server/remote/gitlab/convert.go:76]: it's never checked if err != nil so there is no value in it

[the remark this answers is not in the document]
Comment (6543) [reply server/remote/gitlab/convert.go:76]: @anbraten :eyes:

**Comment by anbraten** [comment]
lgtm
